pretrain/tinyllama.py

```diff
@@ -0,0 +1,340 @@
+                f"iter {metrics['iter']} step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
+                f" {metrics['iter_time'] * 1000:.2f} ms,{' (optimizer.step)' if not is_accumulating else ''}"
+                f" remaining time: {metrics['remaining_time'] / 3600 / 24:.2f} days"
+            )
+
+            throughput_metrics = throughput.compute()
+            metrics.update(throughput_metrics)
+            fabric.log_dict(metrics, step=state["step_count"])
+
+        if val_dataloader is not None and not is_accumulating and state["step_count"] % eval_step_interval == 0:
+            t0 = time.perf_counter()
+            val_loss = validate(fabric, model, val_dataloader)
+            td = time.perf_counter() - t0
+
+            fabric.print(f"iter {state['iter_num']}: val loss {val_loss:.4f}, val time: {td * 1000:.2f} ms")
+            metrics = {
+                "val_loss": val_loss,
+                "total_tokens": model.config.block_size * state["iter_num"] * micro_batch_size * fabric.world_size,
+                "val_ppl": math.exp(val_loss),
+            }
+            fabric.log_dict(metrics, step=state["step_count"])
+            fabric.barrier()
+
+        if not is_accumulating and state["step_count"] % save_step_interval == 0:
+            checkpoint_path = out_dir / f"iter-{state['iter_num']:06d}-ckpt.pth"
+            fabric.print(f"Saving checkpoint to {str(checkpoint_path)!r}")
+            fabric.save(checkpoint_path, state)
+
+
+@torch.no_grad()
+def validate(fabric: L.Fabric, model: nn.Module, val_dataloader: DataLoader) -> float:
+    fabric.print("Validating ...")
+    model.eval()
+
+    losses = torch.zeros(eval_iters, device=fabric.device)
+    for k, val_data in enumerate(val_dataloader):
+        if k >= eval_iters:
+            break
+        input_ids = val_data[:, 0:model.config.block_size].contiguous().long()
+        targets = val_data[:, 1:(model.config.block_size + 1)].contiguous().long()
+        logits = model(input_ids)
+        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        losses[k] = loss
+
+    model.train()
+    return losses.mean().item()
+
+
+def create_dataloaders(fabric: L.Fabric, batch_size: int, block_size: int) -> Tuple[DataLoader, DataLoader]:
+    # Increase by one because we need the next word as well
+    effective_block_size = block_size + 1
+
+    train_datasets = [
+        StreamingDataset(
+            input_dir="data/slimpajama/train",
+            item_loader=TokensLoader(block_size=effective_block_size), 
+            shuffle=True,
+            drop_last=True,
+        ),
+        StreamingDataset(
+            input_dir="data/starcoder",
+            item_loader=TokensLoader(block_size=effective_block_size), 
+            shuffle=True,
+            drop_last=True,
+        ),
+    ]
+
+    # Mix SlimPajama data and Starcoder data with these proportions:
+    weights = (0.693584, 0.306416)
+    weights = [w / sum(weights) for w in weights]
```

Comment:
Not sure what is the purpose of this 🤔. The sum is equal to 1, isn't it?

Reply:
Not if you change the numbers. We can move the weight normalization inside the combined dataset.

Reply:
I think that will make much more sense.